src/test/java/spoon/test/filters/FilterTest.java

```diff
@@ -115,9 +115,17 @@ public void testNameFilter() throws Exception {
 		CtClass<?> foo = factory.Package().get("spoon.test.filters.testclasses").getType("Foo");
 		assertEquals("Foo", foo.getSimpleName());
 		List<CtNamedElement> elements = foo.getElements(new NameFilter<>("i"));
+		// contract: tests NameFilter constructor for cases other than null
 		assertEquals(1, elements.size());
+		// contract: tests NameFilter constructor for null value case
+		assertThrows(IllegalArgumentException.class, () -> foo.getElements(new NameFilter<>(null)));
+		// contract: tests the getType method of the NameFilter class
+		assertEquals("CtNamedElement",new NameFilter<>("i").getType().getSimpleName());
+		// contract: tests the matches method of the NameFilter class
+		assertTrue(new NameFilter<>("i").matches(elements.get(0)));
```

Comment:
This is a self-fulfilling test. The element in `elements` are all elements for which `matches` returned `true`, and so this must be `true`.

Reply:
@slarse I don't get this one, this test is supposed to test the matches method of the class ```NameFilter```, which it is testing. I have tried applying mutation manually like "replacing the method body with false" which the test is killing, so where's the problem?

Reply:
It's not incorrect, it's just redundant. Spoon's filtering mechanism works by gathering all elements for which a filter returns `true`. Thus, by definition, all elements in the filtered list will match the filter that was used. If you look at it in isolation, this is what you have:

```java
List<CtNamedElement> elements = foo.getElements(new NameFilter<>("i"));
// invariant: for every element e in elements, `new NameFilter<>("i") == true` by the definition of Spoon's filters
assertTrue(new NameFilter<>("i").matches(elements.get(0))); // redundant assertion
```

As long as `getElements(Filter)` is working correctly, it's given that every element in `elements` is one such that the used filter's matches method returned true. It is _not_ the responsibility of a test for `NameFilter` to verify that the overall filtering mechanism works as it should, that would fall on a separate test.

Finally, the mutant you are mentioning is killed by the very first assertion (i.e. `assertEquals(1, elements.size())`), because if `matches` always returns `false`, then no elements are collected.

If you in isolation wanted to test the `matches` method of `NameFilter`, you would do something like so instead:

```java
	@Test
	void testNameFilterMatchesReturnsTrueForMatchingName() {
		// contract: NameFilter.matches should return true for an element with a matching name

		// arrange
		String name = "someNiceName";
		NameFilter<CtNamedElement> nameFilter = new NameFilter<>(name);
		CtNamedElement elemWithMatchingName = new Launcher().getFactory().createLocalVariable();
		elemWithMatchingName.setSimpleName(name);
		
		// act
		boolean matches = nameFilter.matches(elemWithMatchingName);
		
		// assert
		assertTrue(matches);
	}
```

Tests should be as precisely controlled as possible, and rely as little as possible on the component they are testing for setting up the preconditions.

Reply:
So I am supposed to follow this structure from now on. I have seen 10 more tests in the codebase following this contract, arrange, act, assert principle I will try to understand those as well.

Actually in most of the tests in spoon contract is missing, AAA is not being followed. But we should follow the best practices while writing the new tests and improving the existing tests. Thanks for letting me know, I appreciate it.

Reply:
> I have seen 10 more tests in the codebase following this contract, arrange, act, assert principle I will try to understand those as well.

You will find that there are hundreds of tests following this principle, but most of them do not contain comments saying which phase is where. I put them there for emphasis, but for most tests they are redundant as it's obvious which part is what.

If you find another test structure that's more appropriate for some given test, feel free to use that. In general, I find that AAA works well and makes for clean and easy to read tests.

> Actually in most of the tests in spoon contract is missing, AAA is not being followed.

Spoon is old and has a lot of legacy from times when things were done differently. You will find a lot of things that aren't great. This is the case for almost any code base you'll ever work on. To paraphrase Uncle Bob, all we can do is to try to leave it in a little bit less of a mess than it was when we got there.